torch/distributed/_functional_collectives.py

```diff
@@ -176,6 +177,10 @@ def all_reduce(self: torch.Tensor, reduceOp: str, group: RANK_TYPES, tag: str =
     return _maybe_wrap_tensor(tensor)
 
 
+def callback(x, group_size, gather_dim):
+    return torch.cat(torch.chunk(x, group_size, dim=0), dim=gather_dim)
```

Comment:
Hmm... is the idea that we want to delay the `cat() and `chunk()` calls until the result of the allgather is actually needed?

We might want to be a bit careful about the (2) compile case I mentioned in the other comment. In particular, could you write a test like this?
```
@torch.compile
def g(x):
    return torch.matmul(x, x)
    
def f(x):
    y = allgather(x)
    return g(y, y)
```

where if you run it with `TORCH_LOGS="aot"`, I think what we'd want is for the `wait_tensor()` + `cat()` + `chunk()` to all be invoked lazily, so they actually show up inside of the graph that compile generates (before the matmul).

Reply:
nit: can you add comment to this function? Thanks.